### PR TITLE
add log messages for buffer fill percent #23253

### PR DESCRIPTION
**Pull request by abesanderson**
Log fill percent on SourceQueue buffer every 10% at info, after 50% at warning.

### CODE REVIEW

**Comment by ktoso** [comment]
Refs #23253

File: akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala [hunk 1]
```diff
@@ -496,8 +496,8 @@ object Source {
    * @param bufferSize size of buffer in element count
    * @param overflowStrategy Strategy that is used when incoming elements cannot fit inside the buffer
    */
-  def queue[T](bufferSize: Int, overflowStrategy: OverflowStrategy): Source[T, SourceQueueWithComplete[T]] =
-    Source.fromGraph(new QueueSource(bufferSize, overflowStrategy).withAttributes(DefaultAttributes.queueSource))
+  def queue[T](bufferSize: Int, overflowStrategy: OverflowStrategy, logBufferFill: Boolean = false): Source[T, SourceQueueWithComplete[T]] =
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala]: this is breaking API and can not be done in such form.

**Review by ktoso — CHANGES_REQUESTED** [review]
I'm not convienced this is a clean way to add logging. This way we'd end up having tens of methods with a boolean for logging... Rather perhaps we could introduce an Attribute called debug logging that could be specifically enabled on sections of a stream (which is better than a global debug flag for those) - though I guess the logging would have to be `val shouldLog = globalDebugEnabled || logDebugAttributePresent`

**Comment by ktoso** [comment]
OK TO TEST

**Comment by patriknw** [comment]
I agree that we must have a better solution. If it's plain debug logging I think it can just be added without any flags, but then there are categories like "verbose debug logging" and warnings. Attributes could be a nice way for streams.

**Comment by abesanderson** [comment]
Adding into attributes for a per-stage flag seems pretty straightforward, not quite sure where to add things for a global debug.  There is a global setting in the materializer, but I don't see any examples where those materializer settings are put into the stage logic.  Maybe a setting for `akka.stream.debug-logging`, though again, I don't really see other examples in the code of passing typesafe config to a stage.  For this case, I think it falls under the category of warnings...just something to give a bit of preliminary idea if buffers are filling.

**Comment by ktoso** [comment]
> . Maybe a setting for akka.stream.debug-logging, though again, I don't really see other examples in the code of passing typesafe config to a stage.

Any stage can get that via: ``

However what we should to is go via attributes I believe, so a stage would get those via `attributes.get[ActorAttributes.VerboseLogging /*thats the new thing to introduce*/]`

File: akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala [hunk 2]
```diff
@@ -36,13 +37,15 @@ import scala.util.control.NonFatal
...
  */
-@InternalApi private[akka] final class QueueSource[T](maxBuffer: Int, overflowStrategy: OverflowStrategy, logBufferFill: Boolean = false) extends GraphStageWithMaterializedValue[SourceShape[T], SourceQueueWithComplete[T]] {
+@InternalApi private[akka] final class QueueSource[T](maxBuffer: Int, overflowStrategy: OverflowStrategy) extends GraphStageWithMaterializedValue[SourceShape[T], SourceQueueWithComplete[T]] {
   import QueueSource._
 
   val out = Outlet[T]("queueSource.out")
   override val shape: SourceShape[T] = SourceShape.of(out)
 
   override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val verboseLogging = inheritedAttributes.contains(VerboseLogging)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala]: exactly 👍

File: akka-stream/src/main/scala/akka/stream/Attributes.scala [hunk 3]
```diff
@@ -227,6 +228,7 @@ object Attributes {
 
   val none: Attributes = Attributes()
 
+  val verboseLogging: Attributes = Attributes(VerboseLogging)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/Attributes.scala]: good, needs some docs on the method as well as on the object.
Docs could be something like "enables verbose logging in stages that chose to support it. Such logging can be useful when debugging unexpected behaviour in element passing, however should not be used in day-to-day use as it likely will result in too vast amounts of logs" or something like that?

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/Attributes.scala]: I think it's fine to put on Attributes btw; Initially was thinking about ActorAttributes but here is fine 👍

**Review by ktoso — COMMENTED** [review]
Looks pretty good now, just needs docs on the object/method and another eye to look at it :)

**Comment by lightbend-cla-validator** [comment]
Hi @xdev-developer,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by abesanderson** [comment]
>< Royally hosed up rebasing to the master.

File: akka-stream/src/main/scala/akka/stream/Attributes.scala (hunk 3 above)
Comment (abesanderson) [inline akka-stream/src/main/scala/akka/stream/Attributes.scala]: Think this description should work....part of what I was adding for was getting buffer information, so wanted to include that note as well.

**Review by patriknw — APPROVED** [review]
LGTM, please rebase

**Review by jrudolph — COMMENTED** [review]
I can see why this would be useful, however, I have two concerns here:

 * It seems ad-hoc to add this kind of logging to a single stage
 * This change currently proposes to log the fill level at INFO level in 10% increments and to escalate to WARNING at 50% full. These numbers seem arbitrary and should be configurable.

Looking at the bigger picture, we should provide ways to observe these kinds of internal state (e.g. as metrics). The actual threshold and triggers for logging (or responding in other ways) should not have to be built into the stages themselves.

Currently lacking a general interface to provide metrics / state information to the use, I wonder if we could provide these values for the queue stage in another way? For the queue, adding API could be quite simple as we already have an external interface that would allow querying the state.

**Comment by abesanderson** [comment]
@jrudolph That's fair enough.  Those cutoffs simply mimicked what was logged on the cluster sharding(another area where a buffer is used instead of backpressure).  To explain where I was coming from: QueueSource seems one of the few recourses when backpressuring is not available, but there is no way(that I'm aware of) to get any sort of health state of the queue.  In the case that manual intervention has to be taken, I didn't have any alerting or notification I was getting into a bad state.  I can easily configure my queue size, which gave me more of a window of time to correct, but no way to see the problem was occurring until it was already too late and dropping messages.  I think the need I saw could be eliminated as well if more of the metrics of stream rates, buffer fill, etc, were available to users.  I do have access to the commercial tools, which provides metrics on actors, but nothing(yet) for streams.

**Comment by patriknw** [comment]
I think @jrudolph has a very valid point. As we discussed lately we should probably try to find more holistic solution to all such metrics, diagnostics, health, ...
That is not something we can request in this PR and will involve considerable work by the core team. Shall we include the logging from this PR in the meantime? Wdyt @jrudolph ?

**Comment by abesanderson** [comment]
It's been quite some time since I started this one, but I believe it may largely be unnecessary with some of the features added since this was created.  If I recall, the situation I was trying to stop was triggered because I had a stream that was feeding into PersistentActor and the cluster sharding was getting slammed with a flood of messages that recovery couldn't keep up with.  The addition of #22638(limit number of concurrent recoveries) helped to mitigate the original issue a lot.  There still may be a need to grab stats from within the QueueSource, but I think that ultimately, it is better solved by backpressuring from the downstream consumers.

**Pull request closed without merging**